### PR TITLE
Issue #4 solved

### PR DESCRIPTION
**Pull request by vedantmalkar**
## Tasks
- [x] Create a new folder: linear_classifiers/
- [x] Implement:
  - `data_utils.py` — load and preprocess CIFAR-10
  - `svm.py` — SVM loss + gradient
  - `softmax.py` — Softmax loss + gradient
  - `optim.py` — SGD optimizer
  - `train.py` — training loop + accuracy calculation
  - `grad_check.py` — numeric gradient validation

- [x] Train on CIFAR-10 (use 10% subset for dev)
- [x] Visualize class weights as images (show learned templates)
- [x] Add a README explaining key observations

*All Tasks have been solved* for issue #4

### CODE REVIEW

File: cv_learn/linear_classifiers/data_utils.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import pickle
+import os
+
+def load_batch(file):
+    with open(file, 'rb') as fo:
+        data_dict = pickle.load(fo, encoding='bytes')
+        X = data_dict[b'data']       #shape is (10000, 3072)
+        Y = data_dict[b'labels']     
+        X = X.reshape(10000, 3, 32, 32).astype("float")
+        Y = np.array(Y)
```
Comment (5usu) [inline cv_learn/linear_classifiers/data_utils.py:16]: be robust to different key encodings and use actual shape rather than 10000: 
X = data_dict[b'data']
Y = data_dict[b'labels']
X = X.reshape(10000, 3, 32, 32).astype("float") With (conceptual):

something like this

File: cv_learn/linear_classifiers/data_utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        temp_y.append(Y) # 5 arrays of array(10000)
+    X_train = np.concatenate(temp_x) # single array of 50k images (shape of (50000,3,32,32)))
+    Y_train = np.concatenate(temp_y)
+    X_test, Y_test = load_batch(os.path.join(root, "test_batch"))
+    print("loaded CIFAR 10")
+    return X_train, Y_train, X_test, Y_test
+
+def preprocess_data(X_train, Y_train, X_test, Y_test):
+    X_train = X_train.reshape(X_train.shape[0], -1).astype(np.float64) #flatten
+    X_test = X_test.reshape(X_test.shape[0], -1).astype(np.float64)
```
Comment (5usu) [inline cv_learn/linear_classifiers/data_utils.py]: Use np.float32 instead of float64 to reduce memory and speed up computations on large datasets

File: cv_learn/linear_classifiers/grad_check.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    print(f"Checking from {num_of_checks} random weights")
+    print("\n-----------------------------------------------------------------------------------------------")
+    print("Index                  Numerical                Analytic                        Relative Error")
+    print("-----------------------------------------------------------------------------------------------")
+
+    for i in range(num_of_checks):
+        ix = tuple([np.random.randint(3073), np.random.randint(10)])            
+
+        grad_analytic_val = grad_analytic[ix]
```
Comment (5usu) [inline cv_learn/linear_classifiers/grad_check.py:19]: use W.shape
replace: ix = tuple([np.random.randint(3073), np.random.randint(10)])
With: ix = (np.random.randint(W.shape[0]), np.random.randint(W.shape[1]))
Reason: generalizes to any W size.

File: cv_learn/linear_classifiers/grad_check.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        print(f"{str(ix)}               {grad_numerical_val}              {grad_analytic_val}              {rel_error}")
+    
+
+
+def main():
+    from data_utils import load_CIFAR_10, preprocess_data
+    
+    print("Initialized CIFAR data set")
+    X_train, y_train, _, _ = load_CIFAR_10('cifar-10-batches-py')
+    X_train, y_train, _, _ = preprocess_data(X_train, y_train, X_train[:10], y_train[:10])
```
Comment (5usu) [inline cv_learn/linear_classifiers/grad_check.py]: Replace: 
X_train, y_train, _, _ = load_CIFAR_10('cifar-10-batches-py')
X_train, y_train, _, _ = preprocess_data(X_train, y_train, X_train[:10], y_train[:10]) 
With:
X_train, y_train, X_test, y_test = load_CIFAR_10('cifar-10-batches-py')
X_train, y_train, X_test, y_test = preprocess_data(X_train, y_train, X_test, y_test)

use the real test and preprocess properly

File: cv_learn/linear_classifiers/optim.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import numpy as np 
+
+def sgd_step(W, dW, learning_rate):
+    W = W - learning_rate * dW      #updates Weights based on gradient
+    return W
```
Comment (5usu) [inline cv_learn/linear_classifiers/optim.py:5]: to avoid allocation 
replace: W -= learning_rate * dW 
this avoids creating another array

File: cv_learn/linear_classifiers/train.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+def accuracy(W,X,y):
+    scores = X.dot(W)
+    predictions = np.argmax(scores, axis=1)
+    return np.mean(predictions == y)            # finds mean of number of correct predictions 
+
+def train(X_train, Y_train, X_val, Y_val, loss_function, learning_rate=0.001, reg=0.0001, num_epochs=20, batch_size=200):
+    num_of_features = 3073
+    num_of_classes = 10
+    W = np.random.randn(num_of_features, num_of_classes) * 0.001
```
Comment (5usu) [inline cv_learn/linear_classifiers/train.py]: replace hard coded shapes 
like, num_of_features = X_train.shape[1]

File: cv_learn/linear_classifiers/train.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    for epoch in range(num_epochs):
+        batches = create_batches(X_train, Y_train, batch_size)
+        epoch_losses = []
+        
+        for X_batch, Y_batch in tqdm(batches, desc=f'Epoch {epoch+1}/{num_epochs}'):
+            loss, dW = loss_function(W, X_batch, Y_batch, reg)
+            W = sgd_step(W, dW, learning_rate)
+            epoch_losses.append(loss)
+        
+        avg_loss = np.mean(epoch_losses)
```
Comment (5usu) [inline cv_learn/linear_classifiers/train.py:35]: use generator-based batching (see optim.py changes) to avoid copying the whole dataset each epoch.

**Review by 5usu — COMMENTED** [review]
make these changes, test locally then commit again

**Comment by vedantmalkar** [comment]
> make these changes, test locally then commit again

changes have been made and committed

**Review by 5usu — APPROVED** [review]

**Comment by 5usu** [comment]
> > make these changes, test locally then commit again
> 
> changes have been made and committed

add images of local testing here

**Comment by vedantmalkar** [comment]
## SVM:
### terminal output:
Screenshot from 2025-10-27 14-08-35

### graphs:
Screenshot from 2025-10-27 14-09-06

### weight visualization: 
Screenshot from 2025-10-27 14-09-18



## Softmax:
### terminal output:
Screenshot from 2025-10-27 14-15-16

### graphs:
Screenshot from 2025-10-27 14-15-30

### weight visualization:
Screenshot from 2025-10-27 14-15-39
